### PR TITLE
Conform `Tensor` to `PointwiseMultiplicative`.

### PR DESCRIPTION
**Pull request by dan-zheng**
Friend PR: https://github.com/apple/swift/pull/25772

### CODE REVIEW

File: Sources/TensorFlow/Core/Tensor.swift [hunk 1]
```diff
@@ -553,6 +551,47 @@ internal extension Tensor where Scalar: TensorFlowFloatingPoint {
...
+extension Tensor: PointwiseMultiplicative where Scalar: Numeric {
+    /// The scalar one tensor.
+    @inlinable
+    public static var one: Tensor { Tensor(1) }
+
+    /// Returns the pointwise reciprocal of `self`.
+    @inlinable
+    public var reciprocal: Tensor { 1 / self }
+
+    /// Multiplies two tensors and produces their product.
```
Comment (rxwei) [inline Sources/TensorFlow/Core/Tensor.swift]:
```suggestion
    /// Multiplies two tensors element-wise and produces their product.
```

File: Sources/TensorFlow/Core/Tensor.swift [hunk 2]
```diff
@@ -553,6 +551,47 @@ internal extension Tensor where Scalar: TensorFlowFloatingPoint {
...
+    /// - Note: `*` supports broadcasting.
+    @inlinable
+    @differentiable(vjp: _vjpMultiply(lhs:rhs:) where Scalar: TensorFlowFloatingPoint)
+    public static func * (lhs: Tensor, rhs: Tensor) -> Tensor {
+        return Raw.mul(lhs, rhs)
+    }
+
+    /// Multiplies two tensors and produces their product.
```
Comment (rxwei) [inline Sources/TensorFlow/Core/Tensor.swift]:
```suggestion
    /// Multiplies two tensors element-wise and produces their product.
```

File: Sources/TensorFlow/Core/Tensor.swift [hunk 3]
```diff
@@ -553,6 +551,47 @@ internal extension Tensor where Scalar: TensorFlowFloatingPoint {
...
+
+    /// Returns the pointwise reciprocal of `self`.
+    @inlinable
+    public var reciprocal: Tensor { 1 / self }
+
+    /// Multiplies two tensors and produces their product.
+    /// - Note: `*` supports broadcasting.
+    @inlinable
+    @differentiable(vjp: _vjpMultiply(lhs:rhs:) where Scalar: TensorFlowFloatingPoint)
+    public static func * (lhs: Tensor, rhs: Tensor) -> Tensor {
```
Comment (rxwei) [inline Sources/TensorFlow/Core/Tensor.swift]: This should be moved where it was, because it is not a protocol requirement and that `PointwiseMultiplicative` may undergo major changes.

Comment (dan-zheng) [reply Sources/TensorFlow/Core/Tensor.swift]: Done.

**Comment by dan-zheng** [comment]
https://github.com/apple/swift/pull/25772 has been merged, so this PR can be merged once CI gets a new toolchain.

**Comment by eaplatanios** [comment]
@dan-zheng Why did you add a reciprocal but not a `./` function to this protocol? Using reciprocal and then multiplication to perform division would be about two times slower than a single `./` call.

**Comment by rxwei** [comment]
`./` should definitely be added as a protocol requirement with a default implementation.

**Comment by dan-zheng** [comment]
> @dan-zheng Why did you add a reciprocal but not a `./` function to this protocol? Using reciprocal and then multiplication to perform division would be about two times slower than a single `./` call.

Good question.

We decided not to add a division operator for mathematical reasons. My understanding:
* Formal ring/[field](https://en.wikipedia.org/wiki/Field_(mathematics)#Classic_definition) definitions typically involve multiplicative inverse, rather than explicit division which is not defined when dividing by zero.
* `Numeric` does not define division: [related discussion](https://forums.swift.org/t/division-related-numeric-extensions/9563).

> `./` should definitely be added as a protocol requirement with a default implementation.

Adding a default implementation (`x .* x.reciprocal`) would cause derived conformances to not be triggered, leaving @eaplatanios's performance concern unaddressed.

**Comment by rxwei** [comment]
> * Formal ring/[field](https://en.wikipedia.org/wiki/Field_(mathematics)#Classic_definition) definitions typically involve multiplicative inverse, rather than explicit division which is not defined when dividing by zero.

The goal shouldn't be to model the exact algebraic structure, but to make these APIs practically useful. Division is by definition multiplication with multiplicative inverse, so it should be added as a default implementation at least. The reason we want to add a protocol requirement (aka. customization point) is that `/` and `/=` can almost always be implemented more efficiently than the default implementation, so we want libraries to be able to customize them.

> * `Numeric` does not define division: [related discussion](https://forums.swift.org/t/division-related-numeric-extensions/9563).

I think this is a different problem. What `PointwiseMultiplicative` really models is a field, for which reason it should not be called `PointwiseMultiplicative`, but we just need a prototype to make optimizers work. If we are to bring this to Swift Evolution, a field protocol is supposed to be a refinement of `Numeric` (if `Numeric` did not have `magnitude`), and use `*` and `/` instead of `.*` and `./`.

In short, this protocol is a prototype that shouldn't be taken super seriously. Ideally, we want to be able to use `Numeric` to represent rings with identity, but `Numeric` annoyingly has a `magnitude` that prevents this.

**Comment by rxwei** [comment]
> Adding a default implementation (`x .* x.reciprocal`) would cause derived conformances to not be triggered, leaving @eaplatanios's performance concern unaddressed.

This is indeed a pain, but it's a compiler limitation/bug. Derived conformances are supposed to use each field's protocol requirement to implement the corresponding parent type's protocol requirement. The derived conformances code path isn't able trigger this automatically yet AFAIK.
